distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerCli.java

```diff
@@ -74,22 +74,19 @@ public void execute(Terminal terminal, OptionSet options, Environment env, Proce
             return;
         }
 
-        validateConfig(options, env);
+        validateConfig(options, origEnv);
 
-        try (KeyStoreWrapper keystore = KeyStoreWrapper.load(env.configFile())) {
-            // setup security
-            final SecureString keystorePassword = getKeystorePassword(keystore, terminal);
-            env = autoConfigureSecurity(terminal, options, processInfo, env, keystorePassword);
-
-            if (keystore != null) {
-                keystore.decrypt(keystorePassword.getChars());
-            }
+        AtomicReference<Environment> env = new AtomicReference<>(origEnv);
 
+        try (var secrets = secureSettingsLoader(env.get()).load(env.get(), terminal, (p) -> {
```

Comment:
This is a little difficult to understand following both the atomic reference and a callback that uses the read password.

Instead, I wonder if the load method could return a small record that includes the secrets and optionalPassword? Then the autoConfigureSecurity call could be within the try as it was before, and env can be overwritten the way it was before.

Reply:
Sounds good, let me try that.